active/0004-extensibility-support-for-deployment-stacks.md

```diff
@@ -0,0 +1,405 @@
+it will be the extension name. The values of the object are extension configurations. Optionality of configuration is
+determined by the extension. If this is compared to a typical resource group deployment, it is aligned because the
+resource group being deployed to is not defined within the deployment itself, but as a parameter to the deployment,
+either through the CLI parameters or as a scope for nested deployments. In this example, the deployment scope(s) are 
+being defined as extension configurations. Expressions in this object can be constrained to expressions that can be 
+compiled to simple directives that both the deployments engine and the stacks service can execute.
+
+Allowed expressions:
+- ARM resource access that can be compiled to a resource ID (`aksCluster` in the example).
+- ARM resource API calls that can be compiled to a URI, HTTP method, and any inputs (`listClusterAdminCredential` in the example).
+- JSON value accessors (`kubeconfig[0].value` in the example).
+- Usage of compile-time constants (the `0` in `kubeconfig[0]`, `namespace: 'default'` in the example).
+
+Disallowed expressions:
+- Usage of deployment parameters.
+- Usage of Bicep or ARM template functions (built-in or user defined).
+- Usage of branched expressions (ternaries).
+
+### The new design from the ARM template perspective
+
+The main Bicep deployment would compile to the ARM template following this paragraph. Only REP relevant data
+is shown. The new "extensionConfigs" deployment property will be similar in style to how deployment parameters and type 
+definitions are defined in deployment templates. Configuration object property values can be literal values such as strings,
+booleans, and objects. Values can also be directives, such as to fetch a value from an ARM resource API. A directive
+consists of one or more evaluation steps. The Bicep compiler will be responsible for converting template language 
+expressions to these directives that both the deployment service can and stacks service can process without need for 
+complex interpretation. The directives will not contain template language expressions to simplify any validation needed
+for acceptable directives.
+
+main.json - The root deployment
+```json
+{
+   "resources": {
+      "extResources": {
+         "type": "Microsoft.Resources/deployments",
+         "apiVersion": "2022-09-01",
+         "name": "extResourcesDeploymentName",
+         "properties": {
+            "extensionConfigs": {
+               "k8s": {
+                  "namespace": {
+                     "type": "string",
+                     "value": "default"
+                  },
+                  "kubeConfig": {
+                     "type": "directive",
+                     "evaluation": [
+                        {
+                           "type": "ArmApiCall",
+                           "resourceId": "/subscriptions/.../resourceGroups/.../Microsoft.ContainerService/managedClusters/...",
```

Comment:
If the resource ID needs to be a compile-time constant, will this feature be usable in multi-cloud deployments? (Maybe it would be usable with Ev2 substitutions, but those aren't available to 3P users.)

Reply:
It's also unclear how this would be derived within a Bicep compilation. Bicep never knows metadata like the subscription ID or RG name about the target scope -- Bicep emits ARM expressions like `subscription().subscriptionId` instead.

Reply:
I've updated this to accept language expressions. Before the nested deployment is submitted, the extensionConfigs language expressions will be evaluated against the parent deployment template. The GET response will return the evaluated copy of this block. To support passing extension configurations down to nested deployments deeper than 1 level, I'm looking into an "inherit" approach where the template engine will substitute an "inherit" directive in the nested deployment properties with the evaluated copy provided to the current deployment before it's PUT.